src/TRestGeant4ParticleSourceCosmics.cxx

```diff
@@ -0,0 +1,114 @@
+
+    std::istringstream iss(particles);
+    std::string name;
+    while (std::getline(iss, name, ',')) {
+        fParticleNames.insert(name);
+    }
+
+    for (const auto& particle : fParticleNames) {
+        if (geant4ParticleNames.find(particle) == geant4ParticleNames.end()) {
+            cerr << "Particle name '" << particle << "' is not allowed." << endl;
+            exit(1);
+        }
+    }
+
+    auto file = TFile::Open(fFilename.c_str(), "READ");
+    if (file == nullptr) {
+        cerr << "File '" << fFilename << "' not found." << endl;
+        exit(1);
+    }
+    for (const auto& particle : fParticleNames) {
+        auto histogram = file->Get<TH2D>(string(particle + "_energy_zenith").c_str());
+        if (histogram == nullptr) {
+            cerr << "Histogram '" << particle << "' not found in file '" << fFilename << "'." << endl;
+            exit(1);
+        }
+        fHistograms[particle] = histogram;
+    }
+
+    double sum = 0;
+    for (const auto& particle : fParticleNames) {
+        auto hist = fHistograms.at(particle);
+        fParticleWeights[particle] += hist->GetEntries();
+        sum += fParticleWeights.at(particle);
+    }
+    for (auto& entry : fParticleWeights) {
+        entry.second /= sum;
+        cout << "TRestGeant4ParticleSourceCosmics::InitFromConfigFile: particle: " << entry.first
+             << " sampling weight: " << entry.second << endl;
+    }
+
+    // TODO: how to safely close the file?
+}
+
+void TRestGeant4ParticleSourceCosmics::Update() {
+    lock_guard<mutex> lock(fMutex);
+
+    RemoveParticles();
+
+    string particleName;
+    if (fParticleNames.size() == 1) {
+        particleName = *fParticleNames.begin();
+    } else {
+        const auto random = fRandom->Uniform(0, 1);
+        // choose a random particle based on the weights
+        double sum = 0;
+        for (const auto& entry : fParticleWeights) {
+            sum += entry.second;
+            if (random <= sum) {
+                particleName = entry.first;
+                break;
+            }
+        }
+    }
+
+    auto& hist = fHistograms.at(particleName);
+
+    double energy, zenith;
+    hist->GetRandom2(energy, zenith, fRandom.get());
+
+    particleName = geant4ParticleNames.at(particleName);
+
+    TRestGeant4Particle particle;
+    particle.SetParticleName(particleName);
+
+    particle.SetEnergy(energy * 1000);  // Convert from MeV to keV
+
+    double phi = fRandom->Uniform(0, 1) * TMath::TwoPi();
+    double zenithRad = zenith * TMath::DegToRad();
+
+    // direction towards -y (can be rotated later)
+    const TVector3 direction = {TMath::Sin(zenithRad) * TMath::Cos(phi), -TMath::Cos(zenithRad),
+                                TMath::Sin(zenithRad) * TMath::Sin(phi)};
+
+    particle.SetDirection(direction);
+
+    AddParticle(particle);
+}
+
+void TRestGeant4ParticleSourceCosmics::SetSeed(unsigned int seed) {
+    cout << "TRestGeant4ParticleSourceCosmics::SetSeed: " << seed << endl;
```

Comment:
Once you assign the `seed` to the random generator perhaps one should update the internal data member, to make sure the value is synced with the class metadata members.

Although perhaps you may recover it from `fRandom`?

Reply:
Yes, you can recover it. In practise it will always be the same as the general geant4 metadata seed.